contracts/cw3-fixed-multisig/src/contract.rs

```diff
@@ -399,5 +457,485 @@ fn calc_range_start(start_after: Option<HumanAddr>) -> Option<Vec<u8>> {
 
 #[cfg(test)]
 mod tests {
-    // TODO
+    use cosmwasm_std::testing::{mock_dependencies, mock_env};
+    use cosmwasm_std::{coin, BankMsg};
+
+    use cw0::Duration;
+    use cw2::{get_contract_version, ContractVersion};
+
+    use crate::msg::Voter;
+
+    use super::*;
+
+    fn mock_env_height<U: Into<HumanAddr>>(sender: U, height: u64) -> Env {
+        let mut env = mock_env(sender, &[]);
+        env.block.height = height;
+        env
+    }
+
+    const OWNER: &str = "admin0001";
+    const VOTER1: &str = "voter0001";
+    const VOTER2: &str = "voter0002";
+    const VOTER3: &str = "voter0003";
+    const SOMEBODY: &str = "somebody";
+
+    fn voter<T: Into<HumanAddr>>(addr: T, weight: u64) -> Voter {
+        Voter {
+            addr: addr.into(),
+            weight,
+        }
+    }
+
+    // this will set up the init for other tests
+    fn setup_test_case<S: Storage, A: Api, Q: Querier>(
+        mut deps: &mut Extern<S, A, Q>,
+        env: Env,
+        required_weight: u64,
+        max_voting_period: Duration,
+    ) -> StdResult<InitResponse<Empty>> {
+        // Init a contract with voters
+        let voters = vec![
+            voter(&env.message.sender, 0),
+            voter(VOTER1, 1),
+            voter(VOTER2, 2),
+            voter(VOTER3, 3),
+        ];
+
+        let init_msg = InitMsg {
+            voters,
+            required_weight,
+            max_voting_period,
+        };
+        init(&mut deps, env, init_msg)
+    }
+
+    #[test]
+    fn test_init_works() {
+        let mut deps = mock_dependencies(20, &[]);
+        let env = mock_env(OWNER, &[]);
+
+        let max_voting_period = Duration::Time(1234567);
+
+        // No voters fails
+        let init_msg = InitMsg {
+            voters: vec![],
+            required_weight: 1,
+            max_voting_period,
+        };
+        let res = init(&mut deps, env.clone(), init_msg);
+
+        // Verify
+        assert!(res.is_err());
+        match res.unwrap_err() {
+            StdError::GenericErr { msg, .. } => assert_eq!(&msg, "No voters"),
+            e => panic!("unexpected error: {}", e),
+        }
+
+        // Zero required weight fails
+        let init_msg = InitMsg {
+            voters: vec![voter(OWNER, 1)],
+            required_weight: 0,
+            max_voting_period,
+        };
+        let res = init(&mut deps, env.clone(), init_msg);
+
+        // Verify
+        assert!(res.is_err());
+        match res.unwrap_err() {
+            StdError::GenericErr { msg, .. } => assert_eq!(&msg, "Required weight cannot be zero"),
+            e => panic!("unexpected error: {}", e),
+        }
+
+        // Zero weights for voters other than sender not allowed
```

Comment:
As above, I would actually allow this case - you convinced me with the "non-voting proposer" concept.